sound/soc/sdw_utils/soc_sdw_utils.c

```diff
@@ -1122,6 +1128,47 @@ int asoc_sdw_parse_sdw_endpoints(struct snd_soc_card *card,
 				if (dai_info->quirk && !(dai_info->quirk & ctx->mc_quirk))
 					continue;
 
+				dlc.dai_name = dai_info->dai_name;
+				codec_dai = snd_soc_find_dai_with_mutex(&dlc);
+				if (!codec_dai) {
+					dev_err(dev, "codec dai %s not registered yet\n",
+						dlc.dai_name);
+					return -EAGAIN;
+				}
+
+				slave = dev_to_sdw_dev(codec_dai->dev);
+				if (!slave)
+					return -EINVAL;
+
+				switch (dai_info->dai_type) {
+				case SOC_SDW_DAI_TYPE_JACK:
+					key = "J"; /* UAJ, RJ, SimpleJack*/
+					break;
+				case SOC_SDW_DAI_TYPE_AMP:
+					key = "Amp"; /* SmartAmp, SimpleAmp*/
+					break;
+				case SOC_SDW_DAI_TYPE_MIC:
+					key = "Mic"; /* SmartMic, SimpleMic, SpeakerMic*/
+					break;
+				default:
+					dev_err(dev, "Invalid dai type %d\n",
+						dai_info->dai_type);
+					return -EINVAL;
+				}
+				for (k = 0; k < slave->sdca_data.num_functions; k++) {
+					dev_dbg(&slave->dev, "function: %s key %s\n", slave->sdca_data.sdca_func[k].name, key);
+					if (strstr(slave->sdca_data.sdca_func[k].name, key)) {
+						dev_dbg(&slave->dev, "Function %s found\n",
+							slave->sdca_data.sdca_func[k].name);
+						break;
+					}
+				}
```

Comment:
Would it be better to use the type here? It gets patched up I think in the case of the older versions of the standard, so it seems a bit nicer than string matching.

Reply:
Yes, my first version is to use type. However, we have different SDCA function types map to the same DAI type.
```c
enum sdca_function_type {
	SDCA_FUNCTION_TYPE_SMART_AMP	= 0x01,	/* Amplifier with protection features */
	SDCA_FUNCTION_TYPE_SIMPLE_AMP	= 0x02,	/* subset of SmartAmp */
	SDCA_FUNCTION_TYPE_SMART_MIC	= 0x03,	/* Smart microphone with acoustic triggers */
	SDCA_FUNCTION_TYPE_SIMPLE_MIC	= 0x04,	/* subset of SmartMic */
	SDCA_FUNCTION_TYPE_SPEAKER_MIC	= 0x05,	/* Combination of SmartMic and SmartAmp */
	SDCA_FUNCTION_TYPE_UAJ		= 0x06,	/* 3.5mm Universal Audio jack */
	SDCA_FUNCTION_TYPE_RJ		= 0x07,	/* Retaskable jack */
	SDCA_FUNCTION_TYPE_SIMPLE_JACK	= 0x08,	/* Subset of UAJ */
	SDCA_FUNCTION_TYPE_HID		= 0x0A,	/* Human Interface Device, for e.g. buttons */
	SDCA_FUNCTION_TYPE_IMP_DEF	= 0x1F,	/* Implementation-defined function */
};
```
For example, SOC_SDW_DAI_TYPE_AMP could be SDCA_FUNCTION_TYPE_SMART_AMP or SDCA_FUNCTION_TYPE_SIMPLE_AMP. And so do SOC_SDW_DAI_TYPE_JACK and SOC_SDW_DAI_TYPE_MIC

Reply:
You could do something like add a function:

```
int asoc_sdw_get_dai_type(u32 type) {
        switch (type) {
                case SDCA_FUNCTION_TYPE_SMART_AMP:
                case SDCA_FUNCTION_TYPE_SIMPLE_AMP:
                        return SOC_SDW_DAI_TYPE_JACK;
                case SDCA_FUNCTION_TYPE_SMART_MIC:
                case SDCA_FUNCTION_TYPE_SIMPLE_MIC:
                case SDCA_FUNCTION_TYPE_SPEAKER_MIC:
                        return SOC_SDW_DAI_TYPE_MIC;
                case SDCA_FUNCTION_TYPE_UAJ:
                case SDCA_FUNCTION_TYPE_RJ:
                case SDCA_FUNCTION_TYPE_SIMPLE_JACK:
                        return SOC_SDW_DAI_TYPE_JACK;
                default:
                        return -EINVAL;
        }
}
```

Then update your loop to something like:

```
for (k = 0; k < slave->sdca_data.num_functions; k++) {
        int dai_type = asoc_sdw_get_dai_type(slave->sdca_data.sdca_func[k].type);

        if (dai_type == dai_info->dai_type) {
                dev_dbg(&slave->dev, "Function %s found\n",
                                slave->sdca_data.sdca_func[k].name);
                break;
        }
}
```

Reply:
Yes, it looks better. Thanks @charleskeepax

Reply:
I tested it by blacklist the codec driver first and then modprobe it after the machine driver probed. And it works.